pkg/infrastructure/nutanix/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,212 @@
+	keyBody := &nutanixclientv3.CategoryKey{
+		APIVersion:  ptr.To("3.1.0"),
+		Description: ptr.To("Openshift Cluster Category Key"),
+		Name:        ptr.To(categoryKey),
+	}
+	respk, err := nutanixCl.V3.CreateOrUpdateCategoryKey(ctx, keyBody)
+	if err != nil {
+		return fmt.Errorf("failed to create the category key %q. %w", categoryKey, err)
+	}
+	logrus.Infof("Created the category key %q", *respk.Name)
+
+	// create the category value "owned".
+	valBody := &nutanixclientv3.CategoryValue{
+		Description: ptr.To("Openshift Cluster Category Value: resources owned by the cluster"),
+		Value:       ptr.To(nutanixtypes.CategoryValueOwned),
+	}
+	respv, err := nutanixCl.V3.CreateOrUpdateCategoryValue(ctx, categoryKey, valBody)
+	if err != nil {
+		return fmt.Errorf("failed to create the category value %q with category key %q. %w", nutanixtypes.CategoryValueOwned, categoryKey, err)
+	}
+	logrus.Infof("Created the category value %q with name %q", *respv.Value, *respv.Name)
+
+	// create the category value "shared".
+	valBody = &nutanixclientv3.CategoryValue{
+		Description: ptr.To("Openshift Cluster Category Value: resources used but not owned by the cluster"),
+		Value:       ptr.To(nutanixtypes.CategoryValueShared),
+	}
+	respv, err = nutanixCl.V3.CreateOrUpdateCategoryValue(ctx, categoryKey, valBody)
+	if err != nil {
+		return fmt.Errorf("failed to create the category value %q with category key %q. %w", nutanixtypes.CategoryValueShared, categoryKey, err)
+	}
+	logrus.Infof("Created the category value %q with name %q", *respv.Value, *respv.Name)
+
+	// upload the rhcos image.
+	imgName := nutanixtypes.RHCOSImageName(in.InfraID)
+	imgURI := string(*in.RhcosImage)
+	imgReq := &nutanixclientv3.ImageIntentInput{}
+	imgSpec := &nutanixclientv3.Image{
+		Name:        &imgName,
+		Description: ptr.To("Created By OpenShift Installer"),
+		Resources: &nutanixclientv3.ImageResources{
+			ImageType: ptr.To("DISK_IMAGE"),
+			SourceURI: &imgURI,
+		},
+	}
+	imgReq.Spec = imgSpec
+	imgMeta := &nutanixclientv3.Metadata{
+		Kind:       ptr.To("image"),
+		Categories: map[string]string{categoryKey: nutanixtypes.CategoryValueOwned},
+	}
+	imgReq.Metadata = imgMeta
+	respi, err := nutanixCl.V3.CreateImage(ctx, imgReq)
+	if err != nil {
+		return fmt.Errorf("failed to create the rhcos image %q. %w", imgName, err)
+	}
+	imgUUID := *respi.Metadata.UUID
+	logrus.Infof("creating the rhcos image %s (uuid: %s).", imgName, imgUUID)
+
+	if taskUUID, ok := respi.Status.ExecutionContext.TaskUUID.(string); ok {
+		logrus.Infof("waiting the image data uploading from %s, taskUUID: %s.", imgURI, taskUUID)
+
+		// Wait till the image creation task is successed.
+		if err = nutanixtypes.WaitForTask(nutanixCl.V3, taskUUID); err != nil {
+			e1 := fmt.Errorf("failed to create the rhcos image %q. %w", imgName, err)
+			logrus.Error(e1)
+			return e1
+		}
+		logrus.Infof("Created and uploaded the rhcos image data %s (uuid: %s).", imgName, imgUUID)
```

Comment:
nit: we don't typically include periods in log messages

Reply:
changed.